File: year2017/day10/solution.lua
local ffi = require('ffi')
ffi.cdef [[
int knot_hash(int* input, int len, int list_size);
const char* full_knot_hash(const char* input, int len);
]]
local clib = ffi.load('solution')

---@param input integer[]
---@param list_size integer?
---@return integer
local function knot_hash(input, list_size)
    list_size = list_size or 256
    local a = ffi.new('int[?]', #input, input)
    return clib.knot_hash(a, #input, list_size)
end

---@param input string
---@return integer
local function full_knot_hash(input)
    input = input .. string.char(17, 31, 73, 47, 23)
    local hash = clib.full_knot_hash(input, #input)
    return ffi.string(hash, 32)
end

local M = {}
M.knot_hash = knot_hash
M.full_knot_hash = full_knot_hash
return M

File: year2017/day10/solution.c
#include <stdio.h>
#include <stdlib.h>
#define LIST_SIZE 256

static void reverse(int* list, int list_size, int len, int p)
{
    int i = p;
    int j = (p + len - 1) % list_size;
    int step = len / 2;

    while (step > 0) {
        int tmp = list[i];
        list[i] = list[j];
        list[j] = tmp;

        i = (i + 1) % list_size;
        j = (j - 1 + list_size) % list_size;
        step -= 1;
    }
}

int knot_hash(int* input, int len, int list_size)
{
    int list[list_size];
    for (int i = 0; i < list_size; i++) {
        list[i] = i;
    }

    int p = 0;  // current position
    int s = 0;  // skip size

    for (int i = 0; i < len; i++) {
        int l = input[i];
        reverse(list, list_size, l, p);
        p = (p + l + s) % list_size;
        s += 1;
    }

    return list[0] * list[1];
}

const char* full_knot_hash(const char* input, int len)
{
    int list[LIST_SIZE];
    for (int i = 0; i < LIST_SIZE; i++) {
        list[i] = i;
    }

    int p = 0;  // current position
    int s = 0;  // skip size

    for (int round = 0; round < 64; round++) {
        for (int i = 0; i < len; i++) {
            int l = input[i];
            reverse(list, LIST_SIZE, l, p);
            p = (p + l + s) % LIST_SIZE;
            s += 1;
        }
    }

    char* hash = (char*)malloc(33);
    for (int block = 0; block < 16; block++) {
        int val = 0;
        for (int i = 0; i < 16; i++) {
            val ^= list[block * 16 + i];
        }
        sprintf(hash + 2 * block, "%02x", val);
    }
    hash[32] = '\0';
    return hash;
}
